src/Polynomials/MonomialBases.jl

```diff
@@ -380,7 +381,13 @@ function _gradient_nd!(
     _gradient_1d!(g,x,orders[d],d)
   end
 
-  z = zero(Mutable(VectorValue{D,T}))
+  #z = zero(Mutable(VectorValue{D,T}))
+  if isbitstype(T)
```

Comment:
Can you check if `z` is inferible? E.g. with `@code_warntype` ?

I would say that this brakes type-stability of this function, unless the compiler is smart enough to take advantage of the `if` statement.

We want to preserve type-stability here. This is a performance-critical point.

Reply:
I have done 
```InteractiveUtils.@code_warntype Gridap.Polynomials._gradient_nd!(v,xi,orders,terms,c,g,Float64)```
and there is no red so I think it is fine. Is that the right way to go about it ?

Reply:
Thanks for running the test. Which were the types of the parameters in the call you tried? I mean the types of `v,xi,orders,terms,c,g`.

Reply:
I used the types of the parameters that were used when that function was called in the MWE [here](https://github.com/gridap/GridapEmbedded.jl/issues/51). Namely:
```
Tv=Gridap.Arrays.CachedVector{Gridap.TensorValues.VectorValue{2, Float64}, Vector{Gridap.TensorValues.VectorValue{2, Float64}}}
Txi=Gridap.TensorValues.VectorValue{2, Float64}
Torders=Tuple{Int64, Int64}
Tterms=Vector{CartesianIndex{2}}
Tc=Gridap.Arrays.CachedMatrix{Float64, Matrix{Float64}}
Tg=Gridap.Arrays.CachedMatrix{Float64, Matrix{Float64}}
```

Reply:
ok, good call. Thus, in your example call `isbitstype(T)` is going to be true, right? as `T=Float64`, and `isbitstype(Float64)` is true. Have you tried what happens with type stability if `isbitstype(T)` is false? On the other hand, why do you need this if else construct? (for my understanding).

Reply:
ok my bad it seems that it is not type stable when i have the dual number arguments ( `isbitstype(T)` false ):

```diff
_gradient_nd!(::Gridap.Arrays.CachedVector{Gridap.TensorValues.VectorValue{2, ReverseDiff.TrackedReal{Float64, Float64, Nothing}}, Vector{Gridap.TensorValues.VectorValue{2, ReverseDiff.TrackedReal{Float64, Float64, Nothing}}}}, ::Gridap.TensorValues.VectorValue{2, ReverseDiff.TrackedReal{Float64, Float64, Nothing}}, ::Tuple{Int64, Int64}, ::Vector{CartesianIndex{2}}, ::Gridap.Arrays.CachedMatrix{ReverseDiff.TrackedReal{Float64, Float64, Nothing}, Matrix{ReverseDiff.TrackedReal{Float64, Float64, Nothing}}}, ::Gridap.Arrays.CachedMatrix{ReverseDiff.TrackedReal{Float64, Float64, Nothing}, Matrix{ReverseDiff.TrackedReal{Float64, Float64, Nothing}}}, ::Type{Float64})
  from _gradient_nd!2(v::AbstractVector{G}, x, orders, terms::AbstractArray{CartesianIndex{D}, 1}, c::AbstractMatrix{T}, g::AbstractMatrix{T}, ::Type{V}) where {G, T, D, V} in Main.StokesCutFEMTests at /home/user/Documents/GitHub/dev/GitHubTestRepo/HelloWorld/tmp.jl:109
Static Parameters
  G = Gridap.TensorValues.VectorValue{2, ReverseDiff.TrackedReal{Float64, Float64, Nothing}}
  T = ReverseDiff.TrackedReal{Float64, Float64, Nothing}
  D = 2
  V = Float64
Arguments
  #self#::Core.Const(Main.StokesCutFEMTests._gradient_nd!2)
  v::Gridap.Arrays.CachedVector{Gridap.TensorValues.VectorValue{2, ReverseDiff.TrackedReal{Float64, Float64, Nothing}}, Vector{Gridap.TensorValues.VectorValue{2, ReverseDiff.TrackedReal{Float64, Float64, Nothing}}}}
  x::Gridap.TensorValues.VectorValue{2, ReverseDiff.TrackedReal{Float64, Float64, Nothing}}
  orders::Tuple{Int64, Int64}
  terms::Vector{CartesianIndex{2}}
  c::Gridap.Arrays.CachedMatrix{ReverseDiff.TrackedReal{Float64, Float64, Nothing}, Matrix{ReverseDiff.TrackedReal{Float64, Float64, Nothing}}}
  g::Gridap.Arrays.CachedMatrix{ReverseDiff.TrackedReal{Float64, Float64, Nothing}, Matrix{ReverseDiff.TrackedReal{Float64, Float64, Nothing}}}
  _::Core.Const(Float64)
Locals
  @_9::Union{Nothing, Tuple{CartesianIndex{2}, Int64}}
  @_10::Union{Nothing, Tuple{Int64, Int64}}
- k::Any
  o::ReverseDiff.TrackedReal{Float64, Float64, Nothing}
  z::Vector{ReverseDiff.TrackedReal{Float64, Float64, Nothing}}
  dim::Int64
  d@_15::Int64
  @_16::Union{Nothing, Tuple{Int64, Int64}}
  @_17::Union{Nothing, Tuple{Int64, Int64}}
  ci::CartesianIndex{2}
  s::Vector{ReverseDiff.TrackedReal{Float64, Float64, Nothing}}
  val@_20::ReverseDiff.TrackedReal{Float64, Float64, Nothing}
  i::Int64
  @_22::Union{Nothing, Tuple{Int64, Int64}}
  q::Int64
  val@_24::ReverseDiff.TrackedReal{Float64, Float64, Nothing}
  val@_25::ReverseDiff.TrackedReal{Float64, Float64, Nothing}
  d@_26::Int64
Body::Nothing
```

for reference the function is defined as:

```julia

function _gradient_nd!(
	v::AbstractVector{G},
	x,
	orders,
	terms::AbstractVector{CartesianIndex{D}},
	c::AbstractMatrix{T},
	g::AbstractMatrix{T},
	::Type{V}) where {G,T,D,V}
      
	dim = D
	for d in 1:dim
	  _evaluate_1d!(c,x,orders[d],d)
	  _gradient_1d!(g,x,orders[d],d)
	end
      
	#z = zero(Mutable(VectorValue{D,T}))
	if isbitstype(T)
	  z = zero(Mutable(VectorValue{D,T}))
	else 
	  z = zeros(T,D)
	end
	
	o = one(T)
	k = 1
      
	for ci in terms
      
	  s = z
	  for i in eachindex(s)
	    @inbounds s[i] = o
	  end
	  for q in 1:dim
	    for d in 1:dim
	      if d != q
		@inbounds s[q] *= c[d,ci[d]]
	      else
		@inbounds s[q] *= g[d,ci[d]]
	      end
	    end
	  end
      
	  k = _set_gradient!(v,s,k,V)
      
  end
      
end
```

I made the amendment to overcome the error which is triggered when running [`@inbounds s[q] *= c[d,ci[d]]`](https://github.com/ConnorMallon/Gridap.jl/blob/shape_derivative/src/Polynomials/MonomialBases.jl#L404):
```setindex!() with non-isbitstype eltype is not supported by StaticArrays. Consider using SizedArray.```.

Reply:
> ok my bad it seems that it is not type stable when i have the dual number arguments ( isbitstype(T) false ):

As far as I understand from the report, it seems that the compiler is able to infer the type of `z`, but not the one of `k`, right? (- k::Any) In such a case, it seems that the issue is in the `set_gradient!` function with the particular types of `v`, `s`, `k`, and `V` with which we call it. Do you understand why it is not able to infer the type of `k` ? BTW, something I think that you can do is to put a breakpoint in the set_gradient! call, and then call @code_warntype with set_gradient! at that point from the debugger console, so that you do not have to cook the function calls. 

Anyway, I think the most frequent case (and the want for which we cannot definitely afford type instabilities) is the case in which T is a Float64 (or more generally, a `Real`), and in this case you observe that the call is type stable right?

> I made the amendment to overcome the error which is triggered when running [@inbounds s[q] *= c[d,ci[d]]](https://github.com/ConnorMallon/Gridap.jl/blob/shape_derivative/src/Polynomials/MonomialBases.jl#L404):
setindex!() with non-isbitstype eltype is not supported by StaticArrays. Consider using SizedArray..

Ok, the triggering of the error message that you mention was introduced in StaticArrays as a result of this discussion: https://github.com/JuliaArrays/StaticArrays.jl/issues/27.  Bottom-line: It is not safe to mutate a mutable static array (i.e., an MArray), when the data type is not isbitstype, i.e., `ReverseDiff.TrackedReal{Float64, Float64, Nothing}` is not isbitstype. BTW, do you know where the `Nothing` in `TrackedReal` comes from? (may be not relevant for the discussion, just for my understanding). 

Thus, the only think that it comes into my mind to fix this is to return a regular dynamic memory array as you are doing now. The only thing is that I would try to avoid allocating the array on each call to `_gradient_nd!`, for performance reasons. You can may be have two different implementations for ` `_gradient_nd!`, one for isbitstypes, and the other for non isbitstypes. The latter has an additional argument with the work array that is allocated once from the calling process. This may involved propagating this into the case of `FieldGradientArray`. This are just thoughs, this idea has to be refined further.

Reply:
> As far as I understand from the report, it seems that the compiler is able to infer the type of `z`, but not the one of `k`, right? (- k::Any) In such a case, it seems that the issue is in the `set_gradient!` function with the particular types of `v`, `s`, `k`, and `V` with which we call it. Do you understand why it is not able to infer the type of `k` ? BTW, something I think that you can do is to put a breakpoint in the set_gradient! call, and then call @code_warntype with set_gradient! at that point from the debugger console, so that you do not have to cook the function calls.

Ok when I do it without cooking it there is no red. So I believe there is type stability in _gradient_nd!. 

`ReverseDiff.TrackedReal{Float64, Float64, Nothing}` is not isbitstype. BTW, do you know where the `Nothing` in `TrackedReal` comes from? (may be not relevant for the discussion, just for my understanding).

It means that we instantiated this object somewhere along the way by the "zero" constructor see [here](https://github.com/JuliaDiff/ReverseDiff.jl/blob/master/src/tracked.jl#L26-L36)

> Thus, the only think that it comes into my mind to fix this is to return a regular dynamic memory array as you are doing now. The only thing is that I would try to avoid allocating the array on each call to `_gradient_nd!`, for performance reasons. You can may be have two different implementations for ` _gradient_nd!`, one for isbitstypes, and the other for non isbitstypes. The latter has an additional argument with the work array that is allocated once from the calling process. This may involved propagating this into the cache of `FieldGradientArray`. This are just thoughs, this idea has to be refined further.

How pressing is it to change the point in which the array is allocated? Is what we have now ok for the moment? What would be the performance difference? It would probably be worth to mention now that I am experiencing pretty slow performance using ReverseDiff in this way at the moment - I am not sure if it is from this or something else.

Reply:
> How pressing is it to change the point in which the array is allocated? Is what we have now ok for the moment? What would be the performance difference? It would probably be worth to mention now that I am experiencing pretty slow performance using ReverseDiff in this way at the moment - I am not sure if it is from this or something else.

It is very hard to answer that questions without empirical evidence (using e.g., a profiler). I would use a profiler to determine where the hot spots are. In regards to the small array dynamic memory allocation, while it might not be the main hotspot, I would say it is more a thing of consistency/perseverance. We have done daunting efforts to avoid this in the innermost kernels of Gridap, and we have evidence that it might be quite harmful for performance on several scenarios. 

Anyway, in this particular case, I would say that the particular instruction flow where this is triggered is quite special, only when you use AD of the fe level set function. Perhaps not a big deal. I do not know what @fverdugo thinks?

Reply:
For profiling, the Julia documentation https://docs.julialang.org/en/v1/manual/profile/ and https://github.com/timholy/ProfileView.jl are your friends. See also the notes I have shared with you on slack.

Reply:
> Ok when I do it without cooking it there is no red. So I believe there is type stability in _gradient_nd!.

For my understanding. Did you identify any difference among your manual call to the function, and the one that you triggered from the debugger at the particular context in which the function is called? I would like to know if there is any cause for the difference.

Reply:
>It means that we instantiated this object somew[here](https://github.com/JuliaDiff/ReverseDiff.jl/blob/master/src/tracked.jl#L26-L36) along the way by the "zero" constructor see here

Ok, thanks for the reference.

Reply:
BTW, I forgot to mention that the VSCode extension for Julia has also some support for integrated profiling https://discourse.julialang.org/t/julia-vs-code-extension-version-v0-17-released/42865

Reply:
> For my understanding. Did you identify any difference among your manual call to the function, and the one that you triggered from the debugger at the particular context in which the function is called? I would like to know if there is any cause for the difference

the difference is i made an error trying to cook the function (did not import all the functions required in my scope). With amendment, there is now no red even when cooking it.

Reply:
ok, perfect. We are on the safe side here, then.  (except for the dynamic memory allocation issue)

Reply:
> Thus, the only think that it comes into my mind to fix this is to return a regular dynamic memory array as you are doing now. The only thing is that I would try to avoid allocating the array on each call to _gradient_nd!, for performance reasons. You can may be have two different implementations for _gradient_nd!, one for isbitstypes, and the other for non isbitstypes. The latter has an additional argument with the work array that is allocated once from the calling process. This may involved propagating this into the cache of FieldGradientArray. This are just thoughs, this idea has to be refined further.

I have made an attempt at this in the most recent commit. Is this the right idea ?

Reply:
> I have made an attempt at this in the most recent commit. Is this the right idea ?

Yes, it is the right idea. Well done. There are still some comments to be addressed. I have put these along with the corresponding LOC.